finat/enriched.py

```diff
@@ -0,0 +1,103 @@
+from __future__ import absolute_import, print_function, division
+from six.moves import map, zip
+
+import gem
+from gem.utils import cached_property
+
+from finat.finiteelementbase import FiniteElementBase
+
+
+class EnrichedElement(FiniteElementBase):
+    """A finite element whose basis functions are the union of the
+    basis functions of several other finite elements."""
+
+    def __init__(self, elements):
+        super(EnrichedElement, self).__init__()
+        self.elements = tuple(elements)
+
+    @cached_property
+    def cell(self):
+        result, = set(elem.cell for elem in self.elements)
+        return result
+
+    @cached_property
+    def degree(self):
+        return tree_map(max, *[elem.degree for elem in self.elements])
+
+    def entity_dofs(self):
+        '''Return the map of topological entities to degrees of
+        freedom for the finite element.'''
+        from FIAT.mixed import concatenate_entity_dofs
+        return concatenate_entity_dofs(self.cell, self.elements)
+
+    def space_dimension(self):
+        '''Return the dimension of the finite element space.'''
+        return sum(elem.space_dimension() for elem in self.elements)
+
+    @cached_property
+    def index_shape(self):
+        return (self.space_dimension(),)
+
+    @cached_property
+    def value_shape(self):
+        '''A tuple indicating the shape of the element.'''
+        shape, = set(elem.value_shape for elem in self.elements)
+        return shape
+
+    def _compose_evaluations(self, results):
+        keys, = set(map(frozenset, results))
+
+        def merge(tables):
+            tables = tuple(tables)
+            zeta = self.get_value_indices()
+            tensors = []
+            for elem, table in zip(self.elements, tables):
+                beta_i = elem.get_indices()
+                tensors.append(gem.ComponentTensor(
+                    gem.Indexed(table, beta_i + zeta),
+                    beta_i
+                ))
```

Comment:
So each sub element provides a `CT(table[beta_i, zeta], beta_i)`, which has free indices zeta (the indices of the outer element)